### PR TITLE
實現TracIn訓練數據影響力評估模組

### LINKED ISSUE #42: 實現TracIn方法來評估訓練數據影響力
# 實現TracIn方法來評估訓練數據影響力

## 功能需求

實現基於論文《Estimating Training Data Influence by Tracing Gradient Descent》(Pang Wei Koh and Percy Liang)提出的TracIn方法，用於評估訓練數據對模型預測與泛化能力的影響。

### 階段1：基礎實現
- [x] 實現基本的TracIn模組，支持檢查點加載和梯度計算
- [x] 實現自影響力(self-influence)計算功能
- [x] 確保與現有訓練流程兼容

### 階段2：擴展功能
- [x] 增強對排序任務的支持(RankingTracInCP)
- [x] 增加對不同損失函數的支持(標準MRL和GHM損失)
- [x] 實現對MacOS MPS設備的支持

### 階段3：泛化能力評估
- [x] 實現訓練樣本對測試樣本影響力的計算
- [x] 優化命令行界面，支持靈活的參數配置
- [x] 整合元數據保存機制

### 階段4：分析與文檔
- [x] 生成詳細的分析報告，解釋計算原理與結果解讀
- [x] 完善README和使用文檔
- [x] 提供案例分析與應用建議

### 預期效益
1. 識別對模型表現有重大影響的訓練樣本
2. 發現模型難以處理的樣本，指導數據增強
3. 提供對模型泛化能力影響的量化分析
4. 指導訓練數據的篩選和加權

### 潛在挑戰
1. 計算資源需求較高，特別是大數據集
2. 需要儲存多個檢查點以獲得準確評估
3. 不同設備(CUDA/MPS/CPU)的兼容性問題

### 驗收標準
1. 能夠成功計算並保存自影響力與影響力分數
2. 支持多種設備和損失函數配置
3. 生成詳盡的分析報告和使用文檔
4. 與現有訓練流程無縫集成

## 參考資料
1. [原始論文](https://arxiv.org/abs/2002.08484)
2. [TracIn GitHub參考實現](https://github.com/frederick0329/TracIn)

### PR DESCRIPTION
**Pull request by sk413025**
實現了TracIn訓練數據影響力評估模組，支持自影響力計算、多設備支持及多種損失函數。詳細內容請參考README_TRACIN.md。

Closes #42

### CODE REVIEW

**Review by sk413025 — COMMENTED** [review]
# TracIn模組實現詳細說明

雖然此PR已合併，但為了讓代碼庫中保留完整記錄，在此添加詳細的實現說明。

## 實現架構

### 代碼結構
```
utils/tracin/
├── __init__.py          # 模組初始化
├── tracin.py            # 基礎TracIn實現
├── ranking_tracin.py    # 排序任務特定實現
└── README.md            # 模組文檔
```

### 核心類
1. **TracInCP** (tracin.py): 
   - 基礎實現，處理檢查點加載和梯度計算
   - 實現了影響力分數計算的核心邏輯
   - 添加了多設備支持(CUDA/MPS/CPU)

2. **RankingTracInCP** (ranking_tracin.py):
   - 繼承自TracInCP，專為排序任務設計
   - 處理成對樣本的梯度計算
   - 支援不同的排序損失函數

3. **命令行工具** (compute_tracin_influence.py):
   - 提供完整的命令行界面
   - 支持多種參數配置
   - 實現了批處理和數據保存機制

## 關鍵技術實現

### 梯度計算方法
```python
def _compute_grad(self, model, x, y, loss_fn):
    """計算單一樣本的梯度"""
    model.zero_grad()
    output = model(x)
    loss = loss_fn(output, y)
    params = [p for p in model.parameters() if p.requires_grad]
    grads = torch.autograd.grad(loss, params)
    return grads

def compute_gradients(self, model, inputs, targets, loss_fn):
    """計算多個樣本的平均梯度"""
    # 實現省略，詳見代碼
```

### 影響力計算邏輯
```python
def compute_influence(self, train_dataset, test_x, test_y, batch_size=1, num_workers=0):
    """計算訓練樣本對測試樣本的影響力"""
    # 使用單個測試樣本和整個訓練數據集計算影響力
    # 實現省略，詳見代碼
```

### 多設備支持
```python
def get_default_device():
    """獲取默認設備，優先順序: CUDA > MPS > CPU"""
    if torch.cuda.is_available():
        return torch.device('cuda')
    elif hasattr(torch.backends, 'mps') and torch.backends.mps.is_available():
        return torch.device('mps')
    else:
        return torch.device('cpu')
```

## 性能優化

### 記憶體優化
- 實現批處理以減少峰值內存使用
- 及時釋放不再需要的梯度張量
- 使用迭代器避免一次加載整個數據集

### 計算效率
- 使用向量化操作代替循環
- 實現了檢查點梯度緩存機制
- 支持多線程數據加載

## 擴展性設計

### 自定義損失函數支持
```python
# 支持標準損失和GHM損失
self.loss_type = loss_type
if self.loss_type == 'standard':
    self.loss_fn = torch.nn.MarginRankingLoss(margin=margin)
elif self.loss_type == 'ghm':
    from losses.ghm_loss import GHMRankingLoss
    self.loss_fn = GHMRankingLoss(bins=ghm_bins, alpha=ghm_alpha)
```

### 多種影響力計算方法
- 自影響力計算: 評估訓練樣本難度
- 交叉影響力計算: 評估訓練樣本對測試樣本的影響
- 支持擴展為其他影響力度量

## 使用場景範例

### 識別困難樣本
```bash
python compute_tracin_influence.py \
    --frequency 500hz \
    --material plastic \
    --checkpoint-dir saved_models/model_checkpoints/your_checkpoint_dir/ \
    --compute-self-influence \
    --loss-type standard
```

### 數據清理與調試
- 使用影響力分數識別標註錯誤的樣本
- 分析不同角度類別的影響力分布差異
- 指導數據增強策略

此PR實現了一個強大而靈活的TracIn模組，為模型訓練和數據分析提供了寶貴的工具。該實現遵循了模塊化設計原則，確保了與現有代碼的無縫集成，同時提供了全面的多設備支持。
